Duct Defaults

bpark1327 opened the pull request:
## Pull Request Description

Allow duct locations/areas to be defaulted if a supply/return `Duct` is provided but has no location/area specified.

New inputs for air distribution systems:
- Required: `HVACDistribution/ConditionedFloorAreaServed`
- Optional: `HVACDistribution/DistributionSystemType/AirDistribution/NumberofReturnRegisters`

## Checklist

Not all may apply:

- [x] EPvalidator.rb has been updated
- [x] Tests (and test files) have been updated
- [x] Documentation has been updated
- [x] `openstudio tasks.rb update_measures` has been run
- [x] No unexpected regression test changes on CI

shorowit commented on tasks.rb:
```diff
@@ -2307,6 +2321,34 @@ def set_hpxml_hvac_distributions(hpxml_file, hpxml)
                                           duct_insulation_r_value: 0,
                                           duct_location: HPXML::LocationAtticUnvented,
                                           duct_surface_area: 50)
+  elsif ['base-misc-defaults.xml'].include? hpxml_file
+    hpxml.hvac_distributions[0].ducts[0].duct_surface_area = nil
+    hpxml.hvac_distributions[0].ducts[1].duct_surface_area = nil
+    hpxml.hvac_distributions[0].ducts[0].duct_location = nil
+    hpxml.hvac_distributions[0].ducts[1].duct_location = nil
+  elsif ['base-misc-defaults3.xml'].include? hpxml_file
+    hpxml.hvac_distributions[0].ducts.each { |duct| duct.duct_surface_area = nil }
+    hpxml.hvac_distributions[1].ducts.each { |duct| duct.duct_surface_area = nil }
+    hpxml.hvac_distributions[4].ducts.each { |duct| duct.duct_surface_area = nil }
+    hpxml.hvac_distributions[5].ducts.each { |duct| duct.duct_surface_area = nil }
+    hpxml.hvac_distributions[0].ducts.each { |duct| duct.duct_location = nil }
+    hpxml.hvac_distributions[1].ducts.each { |duct| duct.duct_location = nil }
+    hpxml.hvac_distributions[4].ducts.each { |duct| duct.duct_location = nil }
+    hpxml.hvac_distributions[5].ducts.each { |duct| duct.duct_location = nil }
+  elsif ['invalid_files/missing-duct-location-and-surface-area.xml'].include? hpxml_file
+    hpxml.hvac_distributions[0].ducts[1].duct_surface_area = nil
+    hpxml.hvac_distributions[0].ducts[1].duct_location = nil
+    hpxml.hvac_distributions[1].ducts[1].duct_surface_area = nil
+    hpxml.hvac_distributions[1].ducts[1].duct_location = nil
+    hpxml.hvac_distributions[4].ducts[1].duct_surface_area = nil
+    hpxml.hvac_distributions[4].ducts[1].duct_location = nil
+    hpxml.hvac_distributions[5].ducts[1].duct_surface_area = nil
+    hpxml.hvac_distributions[5].ducts[1].duct_location = nil
+  elsif ['invalid_files/missing-duct-location.xml'].include? hpxml_file
+    hpxml.hvac_distributions[0].ducts[1].duct_location = nil
+    hpxml.hvac_distributions[1].ducts[1].duct_location = nil
+    hpxml.hvac_distributions[4].ducts[1].duct_location = nil
+    hpxml.hvac_distributions[5].ducts[1].duct_location = nil
```
Surely this can be done more programmatically? Either by looping through all have distributions and checking the type, or looping through hvac distribution indices of interest.

shorowit commented on tasks.rb:
```diff
@@ -58,6 +58,8 @@ def create_hpxmls
     'invalid_files/water-heater-location.xml' => 'base.xml',
     'invalid_files/water-heater-location-other.xml' => 'base.xml',
+    'invalid_files/missing-duct-location-and-surface-area.xml' => 'base-hvac-multiple.xml',
+    'invalid_files/missing-duct-location.xml' => 'base-hvac-multiple.xml',
```
Need a test where total conditioned floor area served > CFA.

bpark1327 commented on HPXMLtoOpenStudio/resources/hvac.rb:
```diff
@@ -4301,4 +4301,48 @@ def self.get_heatpump_temp_assumptions(heat_pump)
+  end
+
+  def self.get_default_duct_locations(hpxml)
+    location_hierarchy = [HPXML::LocationBasementConditioned,
+                          HPXML::LocationBasementUnconditioned,
+                          HPXML::LocationCrawlspaceVented,
+                          HPXML::LocationCrawlspaceUnvented,
+                          HPXML::LocationAtticVented,
+                          HPXML::LocationAtticUnvented,
+                          HPXML::LocationGarage,
+                          HPXML::LocationLivingSpace]
```
I'll put it on hold until we make a decision.

bpark1327 commented on HPXMLtoOpenStudio/resources/hvac.rb:
```diff
@@ -4301,4 +4301,48 @@ def self.get_heatpump_temp_assumptions(heat_pump)
+  def self.get_default_duct_surface_area(duct_type, cfa, ncfl, total_cond_flr_area_served, cond_flr_area_served, num_returns)
+    if duct_type == HPXML::DuctTypeSupply
+      if ncfl == 1
+        return 0.27 * cfa * 1.0 * (cond_flr_area_served / total_cond_flr_area_served)
```
I'll put it on hold until we make a decision.

shorowit replied:
Note that there are two situations that can encounter this:

A. Only ducted systems
2000ft2 CFA house
1. Central AC - 400ft2 CFA served
2. Central AC - 500ft2 CFA served

B. Ducted and ductless systems:
2000ft2 CFA house
1. Room AC
2. Central AC - 400ft2 CFA served
2. Central AC - 500ft2 CFA served

In situation B, it's technically not possible to know if the sum of CFA served is incorrect. So we will not bother comparing sum of CFA served to CFA.

However, we should go with approach #3: Use the distribution system's conditioned floor area served directly in the equation instead of CFA. So the equation becomes, e.g.:
`        return 0.27 * cond_flr_area_served * 1.0`

And I suggest A) calling out `f_out` as a variable, and B) returning two areas (inside/outside or total/outside).

bpark1327 commented on HPXMLtoOpenStudio/measure.rb:
```diff
@@ -501,6 +504,29 @@ def self.set_defaults_and_globals(runner)
+        ducts.each_with_index do |duct, idx|
+          next unless duct.duct_surface_area.nil?
+
+          duct.duct_location = HVAC.get_default_duct_locations(@hpxml)[idx]
+          duct.duct_surface_area = HVAC.get_default_duct_surface_area(duct.duct_type, @cfa, @ncfl, @total_conditioned_floor_area_served, hvac_distribution.conditioned_floor_area_served, hvac_distribution.number_of_return_registers) / ducts.size
```
I'll put it on hold until we make a decision.

shorowit replied:
I think we want the following outcomes (where we show total ducts):

1. Home w/ unconditioned space (garage, attic, crawlspace, uncond basement)

HPXML
- Supply Duct: Area=Outside and Location=unconditioned space
- (2+ stories) Supply Duct: Area=Inside and Location=living space
- Return Duct: Area=Outside and Location=unconditioned space
- (2+ stories) Return Duct: Area=Inside and Location=living space

2. Home w/ conditioned basement

- Supply Duct: Area=Outside and Location=conditioned basement
- (2+ stories) Supply Duct: Area=Inside and Location=living space
- Return Duct: Area=Outside and Location=conditioned basement
- (2+ stories) Return Duct: Area=Inside and Location=living space

3. Home w/o any other spaces

- Supply Duct: Area=Total and Location=living space
- Return Duct: Area=Total and Location=living space

shorowit commented on HPXMLtoOpenStudio/resources/hvac.rb:
```diff
@@ -4301,4 +4301,48 @@ def self.get_heatpump_temp_assumptions(heat_pump)
+      end
+    end
+
+    return space_types
```
Example scenario:

in HPXML:
Duct: Supply, R-4
Duct: Supply, R-8
Duct: Return, R-4
Duct: Return, R-0

out HPXML:
- Supply Duct: Area=Outside and Location=conditioned basement, 50% Area, R-4
- Supply Duct: Area=Outside and Location=conditioned basement, 50% Area, R-8
- (2+ stories) Supply Duct: Area=Inside and Location=living space, 50% Area, R-4
- (2+ stories) Supply Duct: Area=Inside and Location=living space, 50% Area, R-8
- Return Duct: Area=Outside and Location=conditioned basement, 50% Area, R-4
- Return Duct: Area=Outside and Location=conditioned basement, 50% Area, R-0
- (2+ stories) Return Duct: Area=Inside and Location=living space, 50% Area, R-4
- (2+ stories) Return Duct: Area=Inside and Location=living space, 50% Area, R-0

bpark1327 commented on tasks.rb:
```diff
@@ -2820,7 +2823,8 @@ def set_hpxml_hvac_control(hpxml_file, hpxml)
     hpxml.hvac_distributions.add(id: 'HVACDistribution',
-                                 distribution_system_type: HPXML::HVACDistributionTypeAir)
+                                 distribution_system_type: HPXML::HVACDistributionTypeAir,
+                                 conditioned_floor_area_served: 2700)
```
Some files start with "base-foundation-" and "base-enclosure-" are 1,350 ft² and "base-atticroof-conditioned.xml" is 3,600 ft².  These files are changed accordingly.

bpark1327 commented:
> Nice. We're definitely getting close. I made a few changes to the documentation, please look it over and make sure you're comfortable with it. For example, I removed some language that said "the total duct surface areas will be apportioned by the conditioned floor area served by the air distribution system relative to the conditioned floor area of the building", since that seems incorrect based on our latest implementation.
> 
> Also, I decided it was useful in the documentation to refer to _primary_ and _secondary_ duct locations. How about using those terms in the code instead of inside/outside, since it's confusing to have "outside" ducts potentially assigned to living space?

Thanks for editing.  It is easy to understand as well as read well.  I am comfortable with it.  
I agree "inside/outside" terms were somewhat confusing.  I replaced "inside/outside" terms with "primary/secondary" terms in the code.

shorowit commented on HPXMLtoOpenStudio/resources/hvac.rb:
```diff
@@ -3783,4 +3783,41 @@ def self.get_heatpump_temp_assumptions(heat_pump)
+
+  def self.get_default_duct_surface_area(duct_type, ncfl, cfa_served, n_returns)
+    # Fraction of primary ducts (ducts outside conditioned space)
+    f_out = (ncfl == 1) ? 1.0 : 0.75
```
This doesn't look right. Should be ncfl_ag.

shorowit commented on HPXMLtoOpenStudio/resources/hpxml.rb:
```diff
@@ -4201,6 +4222,29 @@ def check_for_errors()
+    air_distributions.each do |dist|
+      heating_systems = dist.hvac_systems.select { |sys| sys if (sys.respond_to? :fraction_heat_load_served) && (sys.fraction_heat_load_served > 0) }
+      cooling_systems = dist.hvac_systems.select { |sys| sys if (sys.respond_to? :fraction_cool_load_served) && (sys.fraction_cool_load_served > 0) }
+      if heating_systems.any? { |sys| sys.distribution_system_idref == dist.id }
```
The ID check seems unnecessary. `dist.hvac_systems` will already only return HVAC systems that are attached to this distribution system.

shorowit commented on HPXMLtoOpenStudio/resources/hpxml.rb:
```diff
@@ -4201,6 +4222,29 @@ def check_for_errors()
+      if heating_systems.any? { |sys| sys.distribution_system_idref == dist.id }
+        heating_dist << dist
+      end
+      if cooling_systems.any? { |sys| sys.distribution_system_idref == dist.id }
```
And here.

shorowit commented on tasks.rb:
```diff
@@ -279,6 +280,8 @@ def create_hpxmls
     'base-misc-defaults.xml' => 'base.xml',
     'base-misc-defaults2.xml' => 'base-dhw-recirc-demand.xml',
+    'base-misc-defaults3.xml' => 'base-hvac-multiple.xml',
+    'base-misc-lighting-none.xml' => 'base.xml',
```
Where did this come from? Remove.

shorowit commented on workflow/sample_files/base-misc-lighting-none.xml:
```diff
@@ -0,0 +1,555 @@
+<?xml version='1.0' encoding='UTF-8'?>
```
Where did this come from? Remove.

shorowit commented on HPXMLtoOpenStudio/tests/test_defaults.rb:
```diff
@@ -0,0 +1,175 @@
+    expected_return_locations = ['basement - conditioned']
+    expected_supply_areas = [729.0]
+    expected_return_areas = [270.0]
+    _test_default_duct_values(hpxml, expected_supply_locations, expected_return_locations, expected_supply_areas, expected_return_areas)
```
Can you add a check on NumberofReturnRegisters too, since it's also optional?

shorowit commented:
Thanks @bpark1327!

shorowit commented:
@joseph-robertson You should be able to use this in the BuildResidentialHPXML measure now.
